src/d3a/models/market/two_sided.py

```diff
@@ -234,49 +254,85 @@ def accept_bid_offer_pair(self, bid, offer, clearing_rate, trade_bid_info, selec
             trade_bid_info = TradeBidOfferInfo(
                 original_bid_rate=original_bid_rate,
-                propagated_bid_rate=bid.price/bid.energy,
-                original_offer_rate=offer.original_offer_price/offer.energy,
-                propagated_offer_rate=offer.price/offer.energy,
+                propagated_bid_rate=market_bid.price/market_bid.energy,
+                original_offer_rate=market_offer.original_offer_price/market_offer.energy,
+                propagated_offer_rate=market_offer.price/market_offer.energy,
                 trade_rate=original_bid_rate)
 
             bid_trade, trade = self.accept_bid_offer_pair(
-                bid, offer, recommended_pair.trade_rate, trade_bid_info, selected_energy
+                market_bid, market_offer, recommended_pair["trade_rate"],
+                trade_bid_info, selected_energy
             )
 
             if trade.residual is not None or bid_trade.residual is not None:
-                recommended_list = self._replace_offers_bids_with_residual_in_matching_list(
-                    recommended_list, index+1, trade, bid_trade
+                recommendations = self._replace_offers_bids_with_residual_in_matching_list(
+                    recommendations, index+1, trade, bid_trade
                 )
 
+    @staticmethod
+    def _validate_requirements_satisfied(
+            bid: Bid, offer: Offer, clearing_rate: float = None,
+            selected_energy: float = None) -> None:
+        """Validate if both trade parties satisfy each other's requirements.
+
+        :raises:
+            InvalidBidOfferPairException: Bid offer pair failed the validation
+        """
+        if ((offer.requirements or bid.requirements) and
+                not RequirementsSatisfiedChecker.is_satisfied(
+                    offer=offer, bid=bid, clearing_rate=clearing_rate,
+                    selected_energy=selected_energy)):
+            # If no requirement dict is satisfied
+            raise InvalidBidOfferPairException(
+                f"OFFER: {offer} & BID: {bid} requirements failed the validation.")
+
+    @classmethod
+    def validate_bid_offer_match(
+            cls, bid: Bid, offer: Offer, clearing_rate: float, selected_energy: float) -> None:
+        """Basic validation function for bid against an offer.
+
+        Raises:
+            InvalidBidOfferPairException: Bid offer pair failed the validation
+        """
+        if not (bid.energy >= selected_energy and
+                offer.energy >= selected_energy and
+                (bid.energy_rate + FLOATING_POINT_TOLERANCE) >= clearing_rate and
+                (bid.energy_rate + FLOATING_POINT_TOLERANCE) >= offer.energy_rate):
+            raise InvalidBidOfferPairException
+        cls._validate_requirements_satisfied(
+            bid=bid, offer=offer, clearing_rate=clearing_rate, selected_energy=selected_energy)
+
     @classmethod
     def _replace_offers_bids_with_residual_in_matching_list(
```

Comment:
A docstring could be helpful here

Reply:
Actually this was here before and I did not pay enough attention to what it does
Now that you mentioned it, I see some room for refactoring/optimizations here and in the `match_recommendations`, but I prefer leaving this to be tackled in [this ticket](https://gridsingularity.atlassian.net/browse/D3ASIM-3532) to not break things and to have enough time to find the best implementation, I'll be sure I address this in the mentioned ticket right after this one is merged, of course if you agree

Reply:
As suggested, following could be added even now
```
"""
If a trade resulted in a residual offer/bid, upcoming matching list needs to be replaced with residual offer/bid
:param matchings: Recomended list of offer/bid matches
:param start_index: Current index of matching
:param offer_trade: Trade info of the successful offer
:param bid_trade: Trade info of the successful bid
:return: The updated matching offer/bid pair list with existing offer/bid replaced with corresponding residual offer/bid
"""
```